src/System.Reflection.MetadataLoadContext/src/System/Reflection/Runtime/BindingFlagSupport/QueryResult.cs

```diff
@@ -0,0 +1,137 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Collections.Generic;
+
+namespace System.Reflection.Runtime.BindingFlagSupport
+{
+    //
+    // Stores the result of a member filtering that's further filtered by the Public, NonPublic, Instance, Static and FlatternHierarchy bits of BindingFlags.
+    // This object is not considered a candidate for long term caching.
+    //
+    // Note: The uninitialized state ("qr = default(QueryResult<M>)) is considered a valid state for this object, and represents an empty list of members.
+    //
+    internal partial struct QueryResult<M> where M : MemberInfo
+    {
+        public QueryResult(BindingFlags bindingAttr, QueriedMemberList<M> queriedMembers)
+        {
+            _lazyCount = 0;
+            _bindingAttr = bindingAttr;
+            _queriedMembers = queriedMembers;
+        }
+
+        public QueryResultEnumerator GetEnumerator() => new QueryResultEnumerator(this);
+
+        /// <summary>
+        /// Returns the number of matching results.
+        /// </summary>
+        public int Count
+        {
+            get
+            {
+                int count = _lazyCount;
+                if (count == 0)
+                {
+                    if (_queriedMembers == null)
+                        return 0;  // This is an uninitialized QueryResult<M>, which is supported and represents a 0-length list of matches.
+
+                    int unfilteredCount = UnfilteredCount;
+                    for (int i = 0; i < unfilteredCount; i++)
+                    {
+                        if (_queriedMembers.Matches(i, _bindingAttr))
+                            count++;
+                    }
+
+                    if (count == 0)
+                    {
+                        // If no matches were found, set ourselves back to the "uninitialized" state so that future 
+                        // calls to Count won't go through this calculation again.
+                        _queriedMembers = null;
+                    }
+
+                    _lazyCount = count;
+
+                }
+                return count;
+            }
+        }
+
+        /// <summary>
+        /// Copies the results to a freshly allocated array. Use this at api boundary points.
+        /// </summary>
+        public M[] ToArray()
+        {
+            int count = Count;
+            if (count == 0)
+                return Array.Empty<M>();
+
+            M[] newArray = new M[count];
+            CopyTo(newArray, 0);
```

Comment:
I might be wrong but there seems to be some redundant/unnecessary work going on here.

We declare an array of size `Count` and to find count we are looping through `UnfilteredCount` `_queriedMembers`. We then call copy which does the exact same loop again. Can this be written more optimally?

Reply:
It does appear at least the calls to .Matches are redundant and methods could be factored to remove that. I'll add this to corert future.